### PR TITLE
Fix bugs in line vert shader

### LINKED ISSUE #5181: Lines with perspective have skewed ends based on sketch apect ratio
```processing
void setup() {
  size(1000, 500, P3D);
}

void draw() {
  hint(ENABLE_STROKE_PERSPECTIVE);
  
  background(0);
  stroke(255);
  strokeWeight(20);

  translate(width/2, height/2);
  line(200, -200,  0, -200,  200,  0);
}
```

Actual:
image

Expected:
image

### LINKED ISSUE #5182: Lines with perspective at zero depth have different thickness than without perspective
```processing
void setup() {
  size(1000, 500, P3D);
}

void draw() {
  hint(ENABLE_STROKE_PERSPECTIVE);
  
  background(0);
  stroke(255);
  strokeWeight(20);

  translate(width/2, height/2);
  line(-200, 0,  0, 200, 0,  0);
}
```

Actual:
image

Expected:
image

### LINKED ISSUE #5183: Lines with one end on eye plane disappear
In the code below, the middle line with end at Z==0 disappears.

```processing
void setup() {
  size(500, 500, P3D);
}

void draw() {
  background(0);
  stroke(255);
  
  // set modelview to identity so that
  // our coordinates go straight into the shader
  setMatrix(new PMatrix3D());
  
  for (int i = -10; i <= 10; i++) {
    line(10 * i,  50, i,
         10 * i, -50, -500);
  }
}
```

Actual:
image

Expected:
image

### LINKED ISSUE #5194: Stroke join disappears when line end is behind the camera
```processing

void setup() {
  size(500, 500, P3D);
}

void draw() {
  
  background(0);

  translate(width/2, height/2);
  
  stroke(255);
  strokeWeight(20);
  noFill();
  
  beginShape();
  
  PGraphicsOpenGL pg = (PGraphicsOpenGL) g;
  
  vertex(0, -200);
  vertex(200, 0);
  vertex(-200, 0, pg.cameraZ+1); // add 1 to place it behind the camera
  
  endShape();
}
```

Actual:
image

Expected:
image

### PR DESCRIPTION
**Pull request by JakubValtar**
Fixes #5181
Fixes #5182
Fixes #5183
Fixes #5194

point vert updated with the code from line vert to keep them in sync

### CODE REVIEW

**Comment by benfry** [comment]
@codeanticode want to review before we merge?

**Comment by benfry** [comment]
Oh look, Github now has a reviewer thing… Don't mind me, I'll just be returning to my hole called 2013.

**Comment by codeanticode** [comment]
ha, no worries, I will take a look now!

**Comment by codeanticode** [comment]
This code is excellent, much cleaner with a nice derivation of the tangent vector that factors out the problematic division by w, and without apparent edge cases. I will run a few tests and will report back, but don't see any side effects so far.

File: core/src/processing/opengl/shaders/LineVert.glsl [hunk 1]
```diff
@@ -34,54 +34,64 @@ attribute vec4 color;
...
+  // to get the right thickness (see https://github.com/processing/processing/issues/5182)
+  // invert Y, projections in Processing invert Y
+  vec2 perspScale = (projectionMatrix * vec4(1, -1, 0, 0)).xy;
+
+  // No Perspective ---
+  // multiply by W (to cancel out division by W later in the pipeline) and
+  // convert from screen to clip (derived from clip to screen above)
+  vec2 noPerspScale = p.w / (0.5 * viewport.zw);
+
+  gl_Position.xy = p.xy + offset.xy * mix(noPerspScale, perspScale, perspective > 0);
```
Comment (codeanticode) [inline core/src/processing/opengl/shaders/LineVert.glsl]: @JakubValtar the glsl compiler on Mac does not automatically convert perspective > 0 into a number, so it gives an error bc does not find a mix(float, float, bool) function. This works:

mix(noPerspScale, perspScale, float(perspective > 0))

Comment (JakubValtar) [inline core/src/processing/opengl/shaders/LineVert.glsl]: Ah, the overload is in GLSL docs, but looks like only for GLSL 4.50. Try to leave there only `perspective` without anything, it is integer of 0 or 1 and should work on it's own.

Comment (JakubValtar) [inline core/src/processing/opengl/shaders/LineVert.glsl]: Or `float(perspective)` if the int does not work either.

**Comment by JakubValtar** [comment]
@codeanticode I modified the shaders and checked that they run with `#version 110` which means that they should run on all cards supporting OpenGL 2.0 or newer. I also modified the version check in shader preprocessor to ignore comments (something I've ran into while testing and which we should have in place).

**Comment by codeanticode** [comment]
@JakubValtar thanks for the last tweaks

@benfry this PR is ready to merge

**Comment by JakubValtar** [comment]
@codeanticode Thanks for the review!

**Comment by JakubValtar** [comment]
Found another one which I accidentally fixed.

Fixes #5194

**Comment by benfry** [comment]
Thanks gentlemen…

**Comment by JakubValtar** [comment]
*tips hat*
